test/images/agnhost/openidmetadata/openidmetadata.go

```diff
@@ -0,0 +1,160 @@
+	fs.StringVar(&tokenPath, "token-path", "", "path to read service account token from")
+	fs.StringVar(&audience, "audience", "", "audience to check on received token")
+	fs.BoolVar(&externalDiscovery, "external-discovery", false, "enable for validating discovery endpoints outside the cluster, this will skip sending a bearer token with requests")
+}
+
+func main(cmd *cobra.Command, args []string) {
+	ctx, err := withOAuth2Client(context.Background())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	raw, err := gettoken()
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Print("OK: Got token")
+	tok, err := jwt.ParseSigned(raw)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var unsafeClaims claims
+	if err := tok.UnsafeClaimsWithoutVerification(&unsafeClaims); err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("OK: got issuer %s", unsafeClaims.Issuer)
+	log.Printf("Full, not-validated claims: \n%#v", unsafeClaims)
+
+	iss, err := oidc.NewProvider(ctx, unsafeClaims.Issuer)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("OK: Constructed OIDC provider for issuer %v", unsafeClaims.Issuer)
+
+	validTok, err := iss.Verifier(&oidc.Config{ClientID: audience}).Verify(ctx, raw)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Print("OK: Validated signature on JWT")
+
+	var safeClaims claims
+	if err := validTok.Claims(&safeClaims); err != nil {
+		log.Fatal(err)
+	}
+	log.Print("OK: Got valid claims from token!")
+	log.Printf("Full, validated claims: \n%#v", &safeClaims)
+}
+
+type kubeName struct {
+	Name string `json:"name"`
+	UID  string `json:"uid"`
+}
+
+type kubeClaims struct {
+	Namespace      string   `json:"namespace"`
+	ServiceAccount kubeName `json:"serviceaccount"`
+}
+
+type claims struct {
+	jwt.Claims
+
+	Kubernetes kubeClaims `json:"kubernetes.io"`
+}
+
+func (k *claims) String() string {
+	return fmt.Sprintf("%s/%s for %s", k.Kubernetes.Namespace, k.Kubernetes.ServiceAccount.Name, k.Audience)
+}
+
+func gettoken() (string, error) {
+	b, err := ioutil.ReadFile(tokenPath)
+	return string(b), err
+}
+
+// withOAuth2Client returns a context that uses the Kubernetes InClusterConfig
+// for HTTP requests.
+// The `oidc` library respects the oauth2.HTTPClient context key; if it is set,
+// the library will use the provided http.Client rather than the default
+// HTTP client.
+// This allows us to ensure requests get routed to the API server, in a client
+// configured with the appropriate CA.
+func withOAuth2Client(context.Context) (context.Context, error) {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(mtaufen): Someday, might want to change this so that we can test that
+	// TokenProjection with an API audience set to the external provider can
+	// make the requests against the external endpoints (in which case we'd send
+	// a different token with a non-Kubernetes audience).
+
+	// If the endpoints are expected to be outside the cluster, do not send the
+	// bearer token.
+	if externalDiscovery {
```

Comment:
Good point. Updated to just call it --in-cluster-discovery, since we don't support authenticated discovery for cluster-external endpoints yet, and switched to just use the default client for external endpoints.